### ISSUE TITLE
what if I write a syncBN from mx.symbol.BatchNorm?

### ISSUE THREAD

**Issue by brucechin** [zhanghang1989/MXNet-Gluon-SyncBN#1]
Hello! we writed a syncBN by modifying mx.symbol.BatchNorm_v1 source code which is depreciated. we'd like to update to mx.symbol.BatchNorm, but it's harder to make it because NNVM is introduced in these new operators.

Maybe we are not able to do it on source code layer. How do you think we can add synchronization into mx.symbol.BN like you've done in gluon.nn.BN?(These two BNs are incompatible so we can not use yours)

**Comment by zhanghang1989** [comment]
I think the best way to implement SyncBN is doing the hacky thing in backend just like caffe implementation https://github.com/yjxiong/caffe/pull/107/files

**Comment by brucechin** [comment]
yeah, we're planning to implement it in the backend using shared memory, which I think is slower than NCCL

**Comment by zhanghang1989** [comment]
Why your previous implementation is not compatible with NNVM?

**Comment by brucechin** [comment]
In mx.symbol.batchnorm_v1, the operator is a class and I can add NCCL communicator/cudaStream into class private variable. they only need to be initialized once.

In mx.symbol.batchnorm, NNVM is introduced and the way how oeprators are registered is changed.  No class is needed in simple operator like BN. Only need to implement BNforward & BNbackward. But I have no idea where to store cudaStream & NCCLcommunicator now.

That's why we plan to implement it using shared memory like you've done in gluon.nn.BN

**Comment by brucechin** [comment]
btw, A deep neural net consists of more than one BN layer at most time. I'm wondering how to assure the same layer in the DNN across different GPUs can share the same SharedTensor in your implementation. If they don't,  mean and variance from different layers in a DNN may be pushed into a SharedTensor?

**Comment by zhanghang1989** [comment]
Each BN layer create its own two SharedTensors during the initialization https://github.com/zhanghang1989/MXNet-Gluon-SyncBN/blob/master/syncbn.py#L145-L146 , which should avoid unexpected sharing.

**Comment by zhanghang1989** [comment]
I really think the solution using backend implementation is the right way to go. Have you tried using `FStatefulCompute` to register the NCCL communicator/cudaStream?

**Comment by kbyran** [comment]
I think the backend should provide the API allowing cross-device operators, in case of reducing efficiency.

**Comment by brucechin** [comment]
I'll try the FStatefulCompute later, thanks for your suggestion.

**Comment by zhanghang1989** [comment]
Thanks @brucechin Please keep me posted how it goes.
@kbyran I agree with you, but it is hard to convince others due to the only SyncBN use case.

**Comment by brucechin** [comment]
@zhanghang1989  Thanks for your help too. We decide to hack the backend using SharedTensor to synchronize mean and variance

**Comment by zhanghang1989** [comment]
Hi @brucechin , good luck to your work. It would be great if you could consider contribute to MXNet when you implemented SynBN in backend.

**Comment by zhanghang1989** [comment]
I used your `SharedTensor` [implementation in backend](https://github.com/brucechin/SharedTensor) and implemented SyncBN in PR https://github.com/apache/incubator-mxnet/pull/11502 
It is working fine with Gluon API, let me know if it works for Symbol API

**Comment by brucechin** [comment]
I‘ve implement it on the backend too and everything is fine

**Comment by brucechin** [comment]
Because I spent the past 2 weeks on my final exams, I just uploaded my implementation both on BatchNorm_v1(located at src/operator/) and BatchNorm(src/operator/nn/).

Besides, my share.h is a very naive implementation. Here is my [commit](https://github.com/brucechin/incubator-mxnet/commit/ae66793d3543aa0e33f0f413257e15bf30be13ab)

**Comment by brucechin** [comment]
It works for symbol API too.

Just check your new sync_batch_norm-inl.h which has improve a lot from my original batch_norm_v1-inl.h(I'm surprised that you found my naive and ugly implementation using sharedTensor in my forked mxnet repo). Thanks for your contribution!

But now src/operator/nn/batch_norm-inl.h has replaced the old batch_norm_v1-inl.h. I think the sync_bn should be based on src/operator/nn/batch_norm-inl.h? But don't worry. Most of the codes for sync can be reused

**Comment by zhanghang1989** [comment]
No problem. Thanks for sharing the implementation of `sharedTensor`, which is really helpful!
I agree that using `src/operator/nn/batch_norm-inl.h ` will probably make the code more efficient.
